platform/darwin/src/MGLOfflineStorage.mm

```diff
@@ -264,6 +264,62 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
     }
 }
 
+#pragma mark Offline merge methods
+
+- (void)addContentsOfFile:(NSString *)filePath withCompletionHandler:(MGLBatchedOfflinePackAdditionCompletionHandler)completion {
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    if (![fileManager isWritableFileAtPath:filePath]) {
+        [NSException raise:NSInvalidArgumentException format:@"The file path must be writable"];
+    }
+    
+    __weak MGLOfflineStorage *weakSelf = self;
+    [self _addContentsOfFile:filePath withCompletionHandler:^(NSError * _Nullable error) {
+        if (error && completion) {
+            completion(error);
+        } else {
+            MGLOfflineStorage *strongSelf = weakSelf;
+            [strongSelf getPacksWithCompletionHandler:^(NSArray<MGLOfflinePack *> *packs, NSError * _Nullable error) {
+                for (MGLOfflinePack *pack in strongSelf.packs) {
+                    [pack invalidate];
+                }
+                strongSelf.packs = [packs mutableCopy];
+                if (completion) {
+                    completion(error);
+                }
+            }];
+        }
+        
+    }];
+}
+
+- (void)addContentsOfURL:(NSURL *)fileURL withCompletionHandler:(MGLBatchedOfflinePackAdditionCompletionHandler)completion {
+    [self addContentsOfFile:fileURL.absoluteString withCompletionHandler:completion];
+}
+
+- (void)_addContentsOfFile:(NSString *)filePath withCompletionHandler:(MGLBatchedOfflinePackAdditionCompletionHandler)completion {
+    self.mbglFileSource->mergeOfflineRegions(std::string(static_cast<const char *>([filePath UTF8String])), [&, completion](mbgl::expected<mbgl::OfflineRegions, std::exception_ptr> result) {
+        NSError *error;
+        NSMutableArray *packs;
+        if (!result) {
+            error = [NSError errorWithDomain:MGLErrorDomain code:-1 userInfo:@{
+                                                                               NSLocalizedDescriptionKey: @(mbgl::util::toString(result.error()).c_str()),
+                                                                               }];
+        } else {
+            auto& regions = result.value();
+            packs = [NSMutableArray arrayWithCapacity:regions.size()];
+            for (auto &region : regions) {
+                MGLOfflinePack *pack = [[MGLOfflinePack alloc] initWithMBGLRegion:new mbgl::OfflineRegion(std::move(region))];
+                [packs addObject:pack];
+            }
+        }
+        if (completion) {
+            dispatch_async(dispatch_get_main_queue(), [&, completion, error, packs](void) {
+                completion(error);
```

Comment:
At this point, we can pass in both `filePath`, and `packs`. Otherwise, all the work above to build up `packs` is for naught.